### ISSUE TITLE
React-page does not work with vite

### ISSUE THREAD

**Issue by enismusliu** [react-page/react-page#1267]
I am facing a problem when trying to use a component that is importing Editor of React Page! 
My component looks like this:
![PageBuliderScreenShot](https://user-images.githubusercontent.com/91558330/214554296-d93e170d-e15d-4632-877d-e5ef6f1085ba.jpg)
But when it tries to load it automatically throws the error:
![PageBuliderScreenShot Error1jpg](https://user-images.githubusercontent.com/91558330/214554544-a12f3b54-888a-48fa-9288-dd494609d67f.jpg)
![image](https://user-images.githubusercontent.com/91558330/214554623-01c38e78-ad7c-4609-89d4-3dcb022a1edd.png)

I am using React v18.2.0 with typescript, and its created with Vite v4.0.4.

Thanks in advance

**Comment by Giraphi** [comment]
Facing the same issue using Vite v4.2.0. Looks like the `<Editor>` component together with at least one cellPlugin (image or slate) doesn't work with Vite. 
Would be great if there'd be any updates on this.

**Comment by macrozone** [comment]
not sure if its related to react-dnd which did some quirky things and maybe an update on that might help.

ideally you check the project out and add a vite example.

Personally I don't see why you would want another runtime then nextjs as such an editor is very often used in a CMS-context where SSR is important. (Nextjs is used in the example and also to create the landingpage. I update that version from time to time and should be very recent)

On the otherhand, we don't do anyhting special and should be a normal react-component, so its worth to find out what the reason is for this problem.

**Comment by andreastoermer** [comment]
@enismusliu @Giraphi Do you find a solution? I've got the same problem now.

**Comment by tedtramonte** [comment]
@macrozone I'm keen to help get this fixed as I'm a huge fan of Vite (and am being blocked by this currently). You've mentioned in this and #1190 that you'd like a Vite example contributed for testing, which I'm happy to do. 

Do you have a preference as to how to include that example? Currently `/example` goes directly to a Next.js app. Should it be restructured to `/example/nextjs` and have `/example/vite`? Are there automated tests that need to be reconfigured to use this path, or updated to build with vite as well?

Happy to move this to a new issue while implementing that so as not to muddy the discussion of this actual issue.

**Comment by macrozone** [comment]
> @macrozone I'm keen to help get this fixed as I'm a huge fan of Vite (and am being blocked by this currently). You've mentioned in this and #1190 that you'd like a Vite example contributed for testing, which I'm happy to do.
> 
> Do you have a preference as to how to include that example? Currently `/example` goes directly to a Next.js app. Should it be restructured to `/example/nextjs` and have `/example/vite`? Are there automated tests that need to be reconfigured to use this path, or updated to build with vite as well?
> 
> Happy to move this to a new issue while implementing that so as not to muddy the discussion of this actual issue.

that would be awesome. this is a good idea to separate them into a nextjs and vite example. 

> Are there automated tests that need to be reconfigured to use this path, or updated to build with vite as well?

not that i am aware off, but the docs and demo also runs off that example. They also double as some kind of smoke-test to check whether it builds properly, that is also why a vite example might be beneficial.

The nextjs will stay the main example, since react-page is often used in content-pages which use ssr. Its also where we can try out new features, e.g. react server-components and other innovations.

**Comment by tedtramonte** [comment]
After implementing the Vite example PR and being unable to reproduce the error, I went back to my real app where I am getting this error and ran the app in "production" which means running `vite build` and serving those files via nginx. To my surprise, the app ran without errors.

Since Vite uses esbuild in dev mode (`vite`) and Rollup in production (`vite build`), it makes me think there's a configuration that could be adjusted in either the esbuild settings to handle the error or in the Rollup settings to reproduce the error. In any case, to me it seems to point to the problem being in the transpilation of some dependency from CommonJS to ESM.

**Comment by natainditama** [comment]
Any updates ?

**Comment by Masab-M** [comment]
Any Updates?

**Comment by zehawki** [comment]
Here's the minimal example where the issue occurs: https://filetransfer.io/data-package/6TlD593O#link

Steps:
1. Copy the files into some folder
2. Run `yarn install`
3. To test Vite is setup properly - comment out the code L1-L24 in src/App.jsx and uncomment L32-L64
4. `yarn dev`
5. Should open http://localhost:3000/ in some browser, and see some hello world type output - all OK
6. Then undo changes in Step 3
7. See error in console:

```
Uncaught Error: Element type is invalid. Received a promise that resolves to: [object Object]. Lazy element type must resolve to a class or function.
    at mountLazyComponent (react-dom.development.js:17733:13)
    at beginWork (react-dom.development.js:19055:16)
    at HTMLUnknownElement.callCallback2 (react-dom.development.js:3945:14)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:16)
    at invokeGuardedCallback (react-dom.development.js:4056:31)
    at beginWork$1 (react-dom.development.js:23964:7)
    at performUnitOfWork (react-dom.development.js:22776:12)
    at workLoopSync (react-dom.development.js:22707:5)
    at renderRootSync (react-dom.development.js:22670:7)
    at performSyncWorkOnRoot (react-dom.development.js:22293:18)
mountLazyComponent @ react-dom.development.js:17733
beginWork @ react-dom.development.js:19055
callCallback2 @ react-dom.development.js:3945
invokeGuardedCallbackDev @ react-dom.development.js:3994
invokeGuardedCallback @ react-dom.development.js:4056
beginWork$1 @ react-dom.development.js:23964
performUnitOfWork @ react-dom.development.js:22776
workLoopSync @ react-dom.development.js:22707
renderRootSync @ react-dom.development.js:22670
performSyncWorkOnRoot @ react-dom.development.js:22293
(anonymous) @ react-dom.development.js:11327
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
flushSyncCallbackQueueImpl @ react-dom.development.js:11322
workLoop @ scheduler.development.js:417
flushWork @ scheduler.development.js:390
performWorkUntilDeadline @ scheduler.development.js:157
Show 17 more frames
Show less
react-dom.development.js:20085 The above error occurred in one of your React components:

    at LazyWithPreload2 (http://localhost:3000/node_modules/.vite/deps/chunk-SKGJGXSK.js?v=1d267225:566:52)
    at Suspense
    at http://localhost:3000/node_modules/.vite/deps/chunk-SKGJGXSK.js?v=1d267225:11066:17
    at div
    at http://localhost:3000/node_modules/.vite/deps/chunk-OW6EROPJ.js?v=1d267225:3548:50
    at Avatar2 (http://localhost:3000/node_modules/.vite/deps/chunk-LEFBF7PF.js?v=1d267225:11289:17)
    at li
    at http://localhost:3000/node_modules/.vite/deps/chunk-OW6EROPJ.js?v=1d267225:3548:50
    at ListItem2 (http://localhost:3000/node_modules/.vite/deps/chunk-LEFBF7PF.js?v=1d267225:19127:17)
    at div
    at Draggable (http://localhost:3000/node_modules/.vite/deps/chunk-SKGJGXSK.js?v=1d267225:12023:20)
    at Item (http://localhost:3000/node_modules/.vite/deps/chunk-SKGJGXSK.js?v=1d267225:12052:19)
    at ul
    at http://localhost:3000/node_modules/.vite/deps/chunk-OW6EROPJ.js?v=1d267225:3548:50
    at List2 (http://localhost:3000/node_modules/.vite/deps/chunk-LEFBF7PF.js?v=1d267225:5681:17)
    at div
    at http://localhost:3000/node_modules/.vite/deps/chunk-OW6EROPJ.js?v=1d267225:3548:50
    at Paper2 (http://localhost:3000/node_modules/.vite/deps/chunk-CJBGWQEM.js?v=1d267225:3983:17)
    at http://localhost:3000/node_modules/.vite/deps/chunk-OW6EROPJ.js?v=1d267225:3548:50
    at Transition2 (http://localhost:3000/node_modules/.vite/deps/chunk-5T76JHYD.js?v=1d267225:86:30)
    at Slide2 (http://localhost:3000/node_modules/.vite/deps/chunk-LEFBF7PF.js?v=1d267225:15026:17)
    at div
    at http://localhost:3000/node_modules/.vite/deps/chunk-OW6EROPJ.js?v=1d267225:3548:50
    at Drawer2 (http://localhost:3000/node_modules/.vite/deps/chunk-LEFBF7PF.js?v=1d267225:15392:17)
    at Portal5 (http://localhost:3000/node_modules/.vite/deps/chunk-SKGJGXSK.js?v=1d267225:11548:5)
    at http://localhost:3000/node_modules/.vite/deps/chunk-SKGJGXSK.js?v=1d267225:12112:16
    at http://localhost:3000/node_modules/.vite/deps/EditableEditor-2IX4OGEK.js?v=1d267225:6383:16
    at div
    at StickyWrapper (http://localhost:3000/node_modules/.vite/deps/EditableEditor-2IX4OGEK.js?v=1d267225:6424:22)
    at div
    at BlurGate (http://localhost:3000/node_modules/.vite/deps/EditableEditor-2IX4OGEK.js?v=1d267225:5982:22)
    at ThemeProvider2 (http://localhost:3000/node_modules/.vite/deps/chunk-OW6EROPJ.js?v=1d267225:6819:5)
    at ThemeProvider3 (http://localhost:3000/node_modules/.vite/deps/chunk-OW6EROPJ.js?v=1d267225:6906:5)
    at ThemeProvider (http://localhost:3000/node_modules/.vite/deps/chunk-4EMJWMXE.js?v=1d267225:141:12)
    at Provider (http://localhost:3000/node_modules/.vite/deps/chunk-SKGJGXSK.js?v=1d267225:7888:20)
    at ReduxProvider (http://localhost:3000/node_modules/.vite/deps/chunk-SKGJGXSK.js?v=1d267225:8636:18)
    at EditorStoreProvider (http://localhost:3000/node_modules/.vite/deps/EditableEditor-2IX4OGEK.js?v=1d267225:6108:22)
    at DndProvider2 (http://localhost:3000/node_modules/.vite/deps/chunk-LEFBF7PF.js?v=1d267225:33658:9)
    at DndProvider2 (http://localhost:3000/node_modules/.vite/deps/EditableEditor-2IX4OGEK.js?v=1d267225:5921:22)
    at CallbacksProvider (http://localhost:3000/node_modules/.vite/deps/EditableEditor-2IX4OGEK.js?v=1d267225:6005:22)
    at RenderOptionsProvider (http://localhost:3000/node_modules/.vite/deps/EditableEditor-2IX4OGEK.js?v=1d267225:6241:22)
    at OptionsProvider (http://localhost:3000/node_modules/.vite/deps/EditableEditor-2IX4OGEK.js?v=1d267225:6203:22)
    at Provider (http://localhost:3000/node_modules/.vite/deps/EditableEditor-2IX4OGEK.js?v=1d267225:6266:16)
    at EditableEditor (http://localhost:3000/node_modules/.vite/deps/EditableEditor-2IX4OGEK.js?v=1d267225:6494:19)
    at LazyWithPreload2 (http://localhost:3000/node_modules/.vite/deps/chunk-SKGJGXSK.js?v=1d267225:566:52)
    at Suspense
    at http://localhost:3000/node_modules/.vite/deps/chunk-SKGJGXSK.js?v=1d267225:11066:17
    at Editor (http://localhost:3000/node_modules/.vite/deps/chunk-TBZS5ICZ.js?v=1d267225:299:15)
    at App (http://localhost:3000/src/App.jsx?t=1701120552498:28:29)

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://reactjs.org/link/error-boundaries to learn more about error boundaries.
logCapturedError @ react-dom.development.js:20085
update.callback @ react-dom.development.js:20118
callCallback @ react-dom.development.js:12318
commitUpdateQueue @ react-dom.development.js:12339
commitLifeCycles @ react-dom.development.js:20736
commitLayoutEffects @ react-dom.development.js:23426
callCallback2 @ react-dom.development.js:3945
invokeGuardedCallbackDev @ react-dom.development.js:3994
invokeGuardedCallback @ react-dom.development.js:4056
commitRootImpl @ react-dom.development.js:23151
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
commitRoot @ react-dom.development.js:22990
performSyncWorkOnRoot @ react-dom.development.js:22329
(anonymous) @ react-dom.development.js:11327
unstable_runWithPriority @ scheduler.development.js:468
runWithPriority$1 @ react-dom.development.js:11276
flushSyncCallbackQueueImpl @ react-dom.development.js:11322
workLoop @ scheduler.development.js:417
flushWork @ scheduler.development.js:390
performWorkUntilDeadline @ scheduler.development.js:157
Show 21 more frames
Show less
react-dom.development.js:17733 Uncaught Error: Element type is invalid. Received a promise that resolves to: [object Object]. Lazy element type must resolve to a class or function.
    at mountLazyComponent (react-dom.development.js:17733:13)
    at beginWork (react-dom.development.js:19055:16)
    at HTMLUnknownElement.callCallback2 (react-dom.development.js:3945:14)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994:16)
    at invokeGuardedCallback (react-dom.development.js:4056:31)
    at beginWork$1 (react-dom.development.js:23964:7)
    at performUnitOfWork (react-dom.development.js:22776:12)
    at workLoopSync (react-dom.development.js:22707:5)
    at renderRootSync (react-dom.development.js:22670:7)
    at performSyncWorkOnRoot (react-dom.development.js:22293:18)
```

**Comment by zehawki** [comment]
@macrozone anything on this? Moving to Vite is quite important for everyone since CRA is as good as dead.

**Comment by macrozone** [comment]
@zehawki there is a vite example in this mr https://github.com/react-page/react-page/pull/1302, but i did not merge it, because it copied too much code and I did not hear any update

however it seems like vite works without problems, as mentioned in that MR. I would appreciate if the vite crowd would continue with that mr and add examples and fix potential problems (if there are any)

**Comment by zehawki** [comment]
@macrozone if you see the setup I've shared, its the basic react-page example (https://react-page.github.io/docs/#/quick-start?id=simple-example) + vite, nothing else at all and the issue occurs.

**Comment by macrozone** [comment]
> @macrozone I think there was a miscommunication about your expectations for that MR and I never got around to revising it. Vite definitely does not work with this project in the field in all cases, but I couldn't reproduce it in the MR. It's still very much a problem, but I had to move on to other work.

so i your mr it worked, but you saying it does not work in the field. So maybe bundling is the problem?

**Comment by macrozone** [comment]
> @macrozone if you see the setup I've shared, its the basic react-page example (https://react-page.github.io/docs/#/quick-start?id=simple-example) + vite, nothing else at all and the issue occurs.

best you take @tedtramonte's MR and check whether you can reproduce the problem! that would be awesome

**Comment by zehawki** [comment]
I'm afraid, I dont understand what you are asking, please elaborate. From what I understand, @tedtramonte provided an *example* to show an issue, and that has touched 118 files. Ive provided an alternative dead simple example to reproduce the same issue. Whats the point of looking at that very complex MR?

**Comment by tedtramonte** [comment]
To be fair, that MR is so bloated because of my misunderstanding, and most of the changed files are simply due to moving the repo structure a bit and duplicating all of the example code. You could still compare only the Vite specific files and come up with something important.

The finer points of JS packaging and bundling are beyond my full understanding, so I can't comment meaningfully on why my MR works but my work project (and many other peoples' projects) do not. I'm happy to try and develop more data points for anyone savvy enough to understand and fix the issue, though.

**Comment by zehawki** [comment]
@tedtramonte - what you did is fine, its just a copy of the example directory and some changes. That's not what I'm talking about - my point is that you created an example in which you are NOT able to reproduce the issue. Ive created one which does. (And it seems the issue is something to do with module export when using lazyload, as found by others - i.e. the resources that I shared)

Further you are using Vite 4, I'm using the latest Vite 5. Your package.json etc has so many packages and dependencies. The example I've provided is bare bone. So we have an example we were looking for to debug the issue.

**Comment by mmendjeli** [comment]
hi there ,
any news on this issue ?

i am using 
"react": "^18.2.0",
"vite": "^4.3.9"

thanks

**Comment by zehawki** [comment]
@macrozone it would be great if you take a look at this - this is deep in the guts of react-page, and I don't even know where to start looking into this. If this package can be built with Vite, I think it would be very useful for the community.

**Comment by macrozone** [comment]
I recommend to add a vite example in a mr, where its easy to reproduce the problem. There was an attempt were there was ambiguous information, it seems that the issue could not be reproduced and it worked with vite.

it would be nice to have a SIMPLE vite example, without duplicated code (as this is a nightmare to review AND maintain), just a simple smoke test example would already be a start.

i unfortunatly due to private  and work constraints I can't work on vite and react-page at the moment and i don't have a running project where i use it. I can try to find someone, but it would really help if the setup would already have been done (With the example)

**Comment by zehawki** [comment]
@macrozone that's already here - https://github.com/react-page/react-page/issues/1267#issuecomment-1828660020

Also this https://github.com/react-page/react-page/issues/1267#issuecomment-1847970207

**Comment by macrozone** [comment]
> @macrozone that's already here - [#1267 (comment)](https://github.com/react-page/react-page/issues/1267#issuecomment-1828660020)
> 
> Also this [#1267 (comment)](https://github.com/react-page/react-page/issues/1267#issuecomment-1847970207)

please add a pull request with the example

**Comment by zehawki** [comment]
Okays, lemme give it a whirl. (Though ideally I dont want to touch the existing /examples since it will break many things, including docs etc)
